app/src/main/java/org/mozilla/fenix/settings/search/AddSearchEngineFragment.kt

```diff
@@ -0,0 +1,282 @@
+        custom_search_engines_learn_more.setOnClickListener {
+            requireContext().let { context ->
+                val intent = SupportUtils.createCustomTabIntent(
+                    context,
+                    SupportUtils.getSumoURLForTopic(
+                        context,
+                        SupportUtils.SumoTopic.CUSTOM_SEARCH_ENGINES
+                    )
+                )
+                startActivity(intent)
+            }
+        }
+    }
+
+    override fun onResume() {
+        super.onResume()
+        (activity as AppCompatActivity).title = getString(R.string.search_engine_add_custom_search_engine_title)
+        (activity as AppCompatActivity).supportActionBar?.show()
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
+        inflater.inflate(R.menu.add_custom_searchengine_menu, menu)
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        return when (item.itemId) {
+            R.id.add_search_engine -> {
+                when (selectedIndex) {
+                    CUSTOM_INDEX -> createCustomEngine()
+                    else -> {
+                        val engine = availableEngines[selectedIndex]
+                        installEngine(engine)
+                    }
+                }
+
+                true
+            }
+            else -> super.onOptionsItemSelected(item)
+        }
+    }
+
+    private fun createCustomEngine() {
+        custom_search_engine_name_field.error = ""
+        custom_search_engine_search_string_field.error = ""
+
+        val name = edit_engine_name.text?.toString() ?: ""
+        val searchString = edit_search_string.text?.toString() ?: ""
+
+        var hasError = false
+        if (name.isEmpty()) {
+            custom_search_engine_name_field.error = resources
+                .getString(R.string.search_add_custom_engine_error_empty_name)
+            hasError = true
+        }
+
+        val existingIdentifiers = requireComponents
+            .search
+            .provider
+            .allSearchEngineIdentifiers()
+            .map { it.toLowerCase(Locale.ROOT) }
+
+        if (existingIdentifiers.contains(name.toLowerCase(Locale.ROOT))) {
+            custom_search_engine_name_field.error = resources
+                .getString(R.string.search_add_custom_engine_error_existing_name, name)
+            hasError = true
+        }
+
+        if (searchString.isEmpty()) {
+            custom_search_engine_search_string_field
+                .error = resources.getString(R.string.search_add_custom_engine_error_empty_search_string)
+            hasError = true
+        }
+
+        if (!searchString.contains("%s")) {
+            custom_search_engine_name_field
+                .error = resources.getString(R.string.search_add_custom_engine_error_missing_template)
+            hasError = true
+        }
+
+        if (hasError) { return }
+
+        viewLifecycleOwner.lifecycleScope.launch {
```

Comment:
nit: remove unneeded second launch
```suggestion
        viewLifecycleOwner.lifecycleScope.launch(Main) {
```

Reply:
I think I need to make sure that this one is using IO?

Reply:
By default I think Main is used. `withContext(IO)` is a better pattern if you want to use the IO thread to compute something.

```kotlin
val result = withContext(IO) { 
  SearchStringValidator.isSearchStringValid(searchString)
}

// ...do stuff with result on main thread
```

Reply:
Also, Colin requested that we always set the dispatcher for launch to make sure `Dispatchers.Default` is never used. 

```lifecycleScope.launch(Main) {
  val result = withContext(IO) { ...IO work }
  
  ...UI thread work
}